iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/CompactionUtils.java

```diff
@@ -493,4 +497,225 @@ public static boolean isDiskHasSpace(double redundancy) {
     }
     return true;
   }
+
+  public static boolean validateSingleTsFileDataCorrectness(TsFileResource resource) {
+    try (TsFileSequenceReader reader = new TsFileSequenceReader(resource.getTsFilePath())) {
+      if (!reader.isComplete()) {
+        logger.error("{} {} illegal tsfile", resource.getTsFilePath(), VALIDATE_FAILED);
+        return false;
+      }
+
+      Map<Long, IChunkMetadata> chunkMetadataMap = getChunkMetadata(reader);
+      if (chunkMetadataMap.isEmpty()) {
+        logger.error("{} there is no data in the file", VALIDATE_FAILED);
+        return false;
+      }
+
+      List<long[]> alignedTimeBatch = new ArrayList<>();
+      reader.position((long) TSFileConfig.MAGIC_STRING.getBytes().length + 1);
+      int pageIndex = 0;
+      byte marker;
+      while ((marker = reader.readMarker()) != MetaMarker.SEPARATOR) {
+        switch (marker) {
+          case MetaMarker.CHUNK_HEADER:
+          case MetaMarker.TIME_CHUNK_HEADER:
+          case MetaMarker.VALUE_CHUNK_HEADER:
+          case MetaMarker.ONLY_ONE_PAGE_CHUNK_HEADER:
+          case MetaMarker.ONLY_ONE_PAGE_TIME_CHUNK_HEADER:
+          case MetaMarker.ONLY_ONE_PAGE_VALUE_CHUNK_HEADER:
+            long chunkOffset = reader.position();
+            ChunkHeader header = reader.readChunkHeader(marker);
+            IChunkMetadata chunkMetadata = chunkMetadataMap.get(chunkOffset - Byte.BYTES);
+            if (!chunkMetadata.getMeasurementUid().equals(header.getMeasurementID())) {
+              logger.error(
+                  "{} chunk start offset is inconsistent with the value in the metadata.",
+                  VALIDATE_FAILED);
+              return false;
+            }
+
+            // empty value chunk
+            int dataSize = header.getDataSize();
+            if (dataSize == 0) {
+              break;
+            }
+
+            boolean isHasStatistic = (header.getChunkType() & 0x3F) == MetaMarker.CHUNK_HEADER;
+            Decoder defaultTimeDecoder =
+                Decoder.getDecoderByType(
+                    TSEncoding.valueOf(TSFileDescriptor.getInstance().getConfig().getTimeEncoder()),
+                    TSDataType.INT64);
```

Comment:
Perhaps it‘s unsafe to use default decoder here if users change time encoder when restarting, can we get the decoder type from the tsfile? But I find TSFileDescriptor.getInstance().getConfig().setTimeEncoder is only used in the test methods, so it's temporarily safe to use like this.